keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/EventController.kt

```diff
@@ -33,22 +34,32 @@ class EventController(
   ): List<ResourceEvent> {
     log.debug("Getting state history for: $id")
     val resource = repository.getResource(id)
-    val pauseScope = resourcePauser.getPauseScope(resource)
-    return repository
-      .resourceEventHistory(id, limit ?: DEFAULT_MAX_EVENTS)
-      .also {
-        if (pauseScope != null) {
-          // for user clarity we add a pause event to the resource history if the resource is paused.
-          val events = it.toMutableList()
-          events.add(
-            0,
-            pausedEvent(resource, "Resource actuation paused at the ${pauseScope.name.toLowerCase()} level")
-          )
-          return events
-        }
+    val events = repository.resourceEventHistory(id, limit ?: DEFAULT_MAX_EVENTS)
+      .toMutableList()
+
+    // For user clarity we add a pause event to the resource history for every pause event from the parent app.
+    // We do this dynamically here so that it applies to all resources in the app, even those added _after_ the
+    // application was paused.
+    val appPausedEvents = repository
+      .applicationEventHistory(resource.application, events.last().timestamp)
+      .filterIsInstance<ApplicationActuationPaused>()
+
+    appPausedEvents.forEach { appPaused ->
+      val lastBeforeAppPaused = events.firstOrNull { event ->
+        event.timestamp.isBefore(appPaused.timestamp)
       }
-  }
 
-  private fun pausedEvent(resource: Resource<*>, message: String) =
-    ResourceActuationPaused(resource, message)
+      if (lastBeforeAppPaused == null) {
+        log.warn("Unable to find a resource event just before application paused event at ${appPaused.timestamp}")
+      } else {
+        events.add(
+          events.indexOf(lastBeforeAppPaused),
+          ResourceActuationPaused(resource, "Resource actuation paused at the application level",
+            appPaused.timestamp)
+        )
+      }
+    }
```

Comment:
This is the gist of the actual fix: we can't just add a singe `ResourceActuationPaused` to the list because they may have been many `ApplicationActuationPaused` events in history, so we add a synthetic event matching each one of those to the list in the right position based on timestamp.

Reply:
Would it make sense to do this in `ActuationPauser` rather than in the controller?

Reply:
Good question. Or maybe just an extension on the `List<ResourceEvent>`?